flink/core/src/main/java/cz/o2/proxima/flink/core/BatchLogSourceFunction.java

```diff
@@ -0,0 +1,336 @@
+                .collect(Collectors.toList());
+        final Set<Partition> skipFirstElement =
+            filteredOffsets
+                .stream()
+                .filter(offset -> offset.getElementIndex() >= 0)
+                .map(Offset::getPartition)
+                .collect(Collectors.toSet());
+        observer = new SourceLogObserver<>(sourceContext, resultExtractor, skipFirstElement);
+        observeHandle =
+            reader.observeOffsets(
+                filteredOffsets, attributeDescriptors, wrapSourceObserver(observer));
+      } else {
+        observer = new SourceLogObserver<>(sourceContext, resultExtractor, Collections.emptySet());
+        observeHandle =
+            reader.observe(partitions, attributeDescriptors, wrapSourceObserver(observer));
+      }
+      log.info("Source [{}]: RUNNING", this);
+      running.countDown();
+      observer.awaitCompleted();
+      final Optional<Throwable> maybeError = observer.getError();
+      if (maybeError.isPresent()) {
+        log.error("Source [{}]: FAILED", this, maybeError.get());
+        ExceptionUtils.rethrowAsIllegalStateException(maybeError.get());
+      } else {
+        finishAndMarkAsIdle(sourceContext);
+      }
+    } else {
+      running.countDown();
+      finishAndMarkAsIdle(sourceContext);
+    }
+  }
+
+  /**
+   * We're done with this source. Progress watermark to {@link Watermark#MAX_WATERMARK}, mark source
+   * as idle and sleep.
+   *
+   * @param sourceContext Source context.
+   * @see <a href="https://issues.apache.org/jira/browse/FLINK-2491">FLINK-2491</a>
+   */
+  @VisibleForTesting
+  void finishAndMarkAsIdle(SourceContext<?> sourceContext) {
+    log.info("Source [{}]: COMPLETED", this);
+    synchronized (sourceContext.getCheckpointLock()) {
+      sourceContext.emitWatermark(new Watermark(Watermarks.MAX_WATERMARK));
+    }
+    sourceContext.markAsTemporarilyIdle();
+  }
+
+  @Override
+  public void cancel() {
+    if (observeHandle != null) {
+      Objects.requireNonNull(observeHandle).close();
+    }
+  }
+
+  @Override
+  public void close() {
+    if (observeHandle != null) {
+      Objects.requireNonNull(observeHandle).close();
+    }
+  }
+
+  protected List<Offset> getCurrentOffsets(ObserveHandle handle) {
+    final OffsetTracking.OffsetTrackingObserveHandle cast =
+        (OffsetTracking.OffsetTrackingObserveHandle) handle;
+    // Filter out finished partitions, as we don't need them for restoring the state.
+    return cast.getCurrentOffsets()
+        .stream()
+        .filter(offset -> !offset.isLast())
+        .collect(Collectors.toList());
+  }
+
+  public void notifyCheckpointComplete(long l) {
+    // No-op.
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throws Exception {
+    Objects.requireNonNull(persistedOffsets).clear();
+    if (observeHandle != null) {
+      for (Offset offset : getCurrentOffsets(Objects.requireNonNull(observeHandle))) {
```

Comment:
redundant `Objects.requireNonNull`

Reply:
For source code analysis, it's not, because object is marked as volatile. That signals it may be changed by another thread at any time.

Reply:
`Optional.ofNullable().ifPresent()` FTW? :)